### PR TITLE
Added `vaultSharesToken` to `IHyperdrive.PoolConfig`

### PR DESCRIPTION
**Pull request by jalextowle**
# Description

This PR adds a new field to `IHyperdrive.PoolConfig` called `vaultSharesToken`. This unifies the interface for getting the vault shares token (something that was previously fragmented across different getters per instance) and also unifies the immutables used by different instances to reduce the complexity of instance implementations. 

# Review Checklists

Please check each item **before approving** the pull request. While going
through the checklist, it is recommended to leave comments on items that are
referenced in the checklist to make sure that they are reviewed. If there are
multiple reviewers, copy the checklists into sections titled `## [Reviewer Name]`.
If the PR doesn't touch Solidity and/or Rust, the corresponding checklist can
be removed.

### Solidity

- [ ] **Tokens**
    - [ ] Do all `approve` calls use `forceApprove`?
    - [ ] Do all `transfer` calls use `safeTransfer`?
    - [ ] Do all `transferFrom` calls use `msg.sender` as the `from` address?
        - [ ] If not, is the function access restricted to prevent unauthorized
              token spend?
- [ ] **Low-level calls (`call`, `delegatecall`, `staticcall`, `transfer`, `send`)**
    - [ ] Is the returned `success` boolean checked to handle failed calls?
    - [ ] If using `delegatecall`, are there strict access controls on the
          addresses that can be called? It shouldn't be possible to `delegatecall`
          arbitrary addresses, so the list of possible targets should either be
          immutable or tightly controlled by an admin.
- [ ] **Reentrancy**
    - [ ] Are functions that make external calls or transfer ether marked as `nonReentrant`?
        - [ ] If not, is there documentation that explains why reentrancy is
              not a concern or how it's mitigated?
- [ ] **Gas Optimizations**
    - [ ] Is the logic as simple as possible?
    - [ ] Are the storage values that are used repeatedly cached in stack or
          memory variables?
    - [ ] If loops are used, are there guards in place to avoid out-of-gas
          issues?
- [ ] **Visibility**
    - [ ] Are all `payable` functions restricted to avoid stuck ether?
- [ ] **Math**
    - [ ] Is all of the arithmetic checked or guarded by if-statements that will
          catch underflows?
    - [ ] If `Safe` functions are altered, are potential underflows and
          overflows caught so that a failure flag can be thrown?
    - [ ] Are all of the rounding directions clearly documented?
- [ ] **Testing**
    - [ ] Are there new or updated unit or integration tests?
    - [ ] Do the tests cover the happy paths?
    - [ ] Do the tests cover the unhappy paths?
    - [ ] Are there an adequate number of fuzz tests to ensure that we are
          covering the full input space?

### Rust

- [ ] **Testing**
    - [ ] Are there new or updated unit or integration tests?
    - [ ] Do the tests cover the happy paths?
    - [ ] Do the tests cover the unhappy paths?
    - [ ] Are there an adequate number of fuzz tests to ensure that we are
          covering the full input space?
    - [ ] If matching Solidity behavior, are there differential fuzz tests that
          ensure that Rust matches Solidity?
          
          
### Solidity

- [x] **Tokens**
    - [x] Do all `approve` calls use `forceApprove`?
    - [x] Do all `transfer` calls use `safeTransfer`?
    - [x] Do all `transferFrom` calls use `msg.sender` as the `from` address?
        - [ ] If not, is the function access restricted to prevent unauthorized
              token spend?
- [x] **Low-level calls (`call`, `delegatecall`, `staticcall`, `transfer`, `send`)**
    - [x] Is the returned `success` boolean checked to handle failed calls?
    - [x] If using `delegatecall`, are there strict access controls on the
          addresses that can be called? It shouldn't be possible to `delegatecall`
          arbitrary addresses, so the list of possible targets should either be
          immutable or tightly controlled by an admin.
- [x] **Reentrancy**
    - [x] Are functions that make external calls or transfer ether marked as `nonReentrant`?
        - [x] If not, is there documentation that explains why reentrancy is
              not a concern or how it's mitigated?
- [x] **Gas Optimizations**
    - [x] Is the logic as simple as possible?
    - [x] Are the storage values that are used repeatedly cached in stack or
          memory variables?
    - [x] If loops are used, are there guards in place to avoid out-of-gas
          issues?
- [x] **Visibility**
    - [x] Are all `payable` functions restricted to avoid stuck ether?
- [x] **Math**
    - [x] Is all of the arithmetic checked or guarded by if-statements that will
          catch underflows?
    - [x] If `Safe` functions are altered, are potential underflows and
          overflows caught so that a failure flag can be thrown?
    - [x] Are all of the rounding directions clearly documented?
- [x] **Testing**
    - [x] Are there new or updated unit or integration tests?
    - [x] Do the tests cover the happy paths?
    - [x] Do the tests cover the unhappy paths?
    - [x] Are there an adequate number of fuzz tests to ensure that we are
          covering the full input space?

### Rust

- [ ] **Testing**
    - [ ] Are there new or updated unit or integration tests?
    - [ ] Do the tests cover the happy paths?
    - [ ] Do the tests cover the unhappy paths?
    - [ ] Are there an adequate number of fuzz tests to ensure that we are
          covering the full input space?
    - [ ] If matching Solidity behavior, are there differential fuzz tests that
          ensure that Rust matches Solidity?
          

### Solidity

- [x] **Tokens**
    - [x] Do all `approve` calls use `forceApprove`?
    - [x] Do all `transfer` calls use `safeTransfer`?
    - [x] Do all `transferFrom` calls use `msg.sender` as the `from` address?
        - [x] If not, is the function access restricted to prevent unauthorized
              token spend?
- [x] **Low-level calls (`call`, `delegatecall`, `staticcall`, `transfer`, `send`)**
    - [x] Is the returned `success` boolean checked to handle failed calls?
    - [x] If using `delegatecall`, are there strict access controls on the
          addresses that can be called? It shouldn't be possible to `delegatecall`
          arbitrary addresses, so the list of possible targets should either be
          immutable or tightly controlled by an admin.
- [x] **Reentrancy**
    - [x] Are functions that make external calls or transfer ether marked as `nonReentrant`?
        - [x] If not, is there documentation that explains why reentrancy is
              not a concern or how it's mitigated?
- [x] **Gas Optimizations**
    - [x] Is the logic as simple as possible?
    - [x] Are the storage values that are used repeatedly cached in stack or
          memory variables?
    - [x] If loops are used, are there guards in place to avoid out-of-gas
          issues?
- [x] **Visibility**
    - [x] Are all `payable` functions restricted to avoid stuck ether?
- [x] **Math**
    - [x] Is all of the arithmetic checked or guarded by if-statements that will
          catch underflows?
    - [x] If `Safe` functions are altered, are potential underflows and
          overflows caught so that a failure flag can be thrown?
    - [x] Are all of the rounding directions clearly documented?
- [x] **Testing**
    - [x] Are there new or updated unit or integration tests?
    - [x] Do the tests cover the happy paths?
    - [x] Do the tests cover the unhappy paths?
    - [x] Are there an adequate number of fuzz tests to ensure that we are
          covering the full input space?

### Rust

- [ ] **Testing**
    - [ ] Are there new or updated unit or integration tests?
    - [ ] Do the tests cover the happy paths?
    - [ ] Do the tests cover the unhappy paths?
    - [ ] Are there an adequate number of fuzz tests to ensure that we are
          covering the full input space?
    - [ ] If matching Solidity behavior, are there differential fuzz tests that
          ensure that Rust matches Solidity?

### Solidity

- [ ] **Tokens**
    - [x] Do all `approve` calls use `forceApprove`?
    - [ ] Do all `transfer` calls use `safeTransfer`?
    - [x] Do all `transferFrom` calls use `msg.sender` as the `from` address?
        - [x] If not, is the function access restricted to prevent unauthorized
              token spend?
- [x] **Low-level calls (`call`, `delegatecall`, `staticcall`, `transfer`, `send`)**
    - [x] Is the returned `success` boolean checked to handle failed calls?
    - [x] If using `delegatecall`, are there strict access controls on the
          addresses that can be called? It shouldn't be possible to `delegatecall`
          arbitrary addresses, so the list of possible targets should either be
          immutable or tightly controlled by an admin.
- [x] **Reentrancy**
    - [x] Are functions that make external calls or transfer ether marked as `nonReentrant`?
        - [x] If not, is there documentation that explains why reentrancy is
              not a concern or how it's mitigated?
- [x] **Gas Optimizations**
    - [x] Is the logic as simple as possible?
    - [x] Are the storage values that are used repeatedly cached in stack or
          memory variables?
    - [x] If loops are used, are there guards in place to avoid out-of-gas
          issues?
- [x] **Visibility**
    - [x] Are all `payable` functions restricted to avoid stuck ether?
- [x] **Math**
    - [x] Is all of the arithmetic checked or guarded by if-statements that will
          catch underflows?
    - [x] If `Safe` functions are altered, are potential underflows and
          overflows caught so that a failure flag can be thrown?
    - [x] Are all of the rounding directions clearly documented?
- [x] **Testing**
    - [x] Are there new or updated unit or integration tests?
    - [x] Do the tests cover the happy paths?
    - [x] Do the tests cover the unhappy paths?
    - [x] Are there an adequate number of fuzz tests to ensure that we are
          covering the full input space?

### CODE REVIEW

File: contracts/src/instances/erc4626/ERC4626Hyperdrive.sol
Comment (jalextowle) [inline contracts/src/instances/erc4626/ERC4626Hyperdrive.sol:87]: All of the assertions in the instances were moved to the deployer coordinator for consistency and to reduce the code size.

File: contracts/src/instances/reth/RETHBase.sol
Comment (sentilesdal) [inline contracts/src/instances/reth/RETHBase.sol:49]: no safeTransferFrom available.

File: contracts/src/instances/lseth/LsETHBase.sol
Comment (sentilesdal) [inline contracts/src/instances/lseth/LsETHBase.sol:36]: no safeTransferFrom available.

File: test/integrations/factory/HyperdriveFactory.t.sol
Comment (sentilesdal) [inline test/integrations/factory/HyperdriveFactory.t.sol:2366]: Address this before merging?

**Review by sentilesdal — COMMENTED** [review]
Went through the checklist, looks mostly mechanincal though.  I double-checked that the integration tokens don't have `safeTranfserFrom` and left comments.  We might want to add a comment to those tokens to make future reviews easier.

After this lands I'll go through the codegen again and update to these changes.

File: contracts/src/instances/erc4626/ERC4626Hyperdrive.sol
Comment (mcclurejt) [inline contracts/src/instances/erc4626/ERC4626Hyperdrive.sol:77]: question: why do this only for ERC4626 yield sources?

**Review by mcclurejt — APPROVED** [review]
LGTM, make sure to address [FIXME](https://github.com/delvtech/hyperdrive/pull/892/files#r1541118696) if needed

File: contracts/src/instances/erc4626/ERC4626Hyperdrive.sol
Comment (jalextowle) [reply contracts/src/instances/erc4626/ERC4626Hyperdrive.sol:77]: The `ERC4626Hyperdrive` instance is the only integration that supports a large number of tokens. All of the other integrations are 1:1 with the underlying token (for example, `StETHHyperdrive` is always `ETH` and `Lido`). Furthermore, all of the other integrations use `ETH` as their base token, so they don't even need to deal with approvals since they can directly forward ETH.

With this in mind, we also don't need to use `forceApprove` and `safeTransferFrom` since we know that the vaults are ERC20 compliant (we only need these libraries for implementations that aren't ERC20 compliant). With this said, I think we should use `safeTransferFrom` and `forceApprove` as a general policy to cover our ass and add some layers of security.

**Review by cashd — APPROVED** [review]

File: contracts/src/deployers/HyperdriveDeployerCoordinator.sol
Comment (jrhea) [inline contracts/src/deployers/HyperdriveDeployerCoordinator.sol:442]:
```suggestion
    /// @param _deployConfig The deploy config that will be used to deploy the
```

File: contracts/src/deployers/erc4626/ERC4626HyperdriveDeployerCoordinator.sol
Comment (jrhea) [inline contracts/src/deployers/erc4626/ERC4626HyperdriveDeployerCoordinator.sol:137]:
```suggestion
    /// @param _deployConfig The deploy config that will be used to deploy the
```

File: contracts/src/external/HyperdriveTarget0.sol
Comment (jrhea) [inline contracts/src/external/HyperdriveTarget0.sol:246]:
```suggestion
    /// @return The vault shares token address.
```

File: contracts/src/interfaces/IERC4626.sol
Comment (jrhea) [inline contracts/src/interfaces/IERC4626.sol:61]: missing owner natspec

**Review by jrhea — APPROVED** [review]
im going to go ahead and approve, but there are some nits to address. also, i don't like that i cant see the coverage changes - why is it skipped in CI?

**Pull request merged**
